super_editor/lib/src/infrastructure/text_layout.dart

```diff
@@ -0,0 +1,96 @@
+import 'package:flutter/rendering.dart';
+
+/// Contract to interrogate the layout of a blob of text.
+abstract class TextLayout {
+  /// Returns [true] if a text character overlaps the given [localOffset],
+  /// or [false] otherwise.
+  bool isTextAtOffset(Offset localOffset);
+
+  /// Returns the [TextPosition] that overlaps the given [localOffset].
+  TextPosition? getPositionAtOffset(Offset localOffset);
+
+  /// Returns the [TextPosition] that overlaps the given [localOffset],
+  /// or the [TextPosition] that is nearest the given [localOffset] if
+  /// no [TextPosition] overlaps the given [localOffset].
+  TextPosition getPositionNearestToOffset(Offset localOffset);
+
+  /// Returns the [Offset] of the character at the given [position].
+  Offset getOffsetAtPosition(TextPosition position);
+
+  /// Returns a [List] of [TextBox]es that contain the given [selection].
+  List<TextBox> getBoxesForSelection(TextSelection selection);
+
+  /// Returns a bounding [Rect] for the character at the given [position].
+  Rect getCharacterBox(TextPosition position);
```

Comment:
I would rename this as it gives an impression to return `TextBox` and not a `Rect`

```suggestion
  Rect getCharacterRect(TextPosition position);
```

Reply:
Good point. I'm gonna change the return type to `TextBox` to give it symmetry with `getBoxesForSelection()`.